src/test/java/jalf/optimizer/TestOptimizedUnion.java

```diff
@@ -0,0 +1,58 @@
+package jalf.optimizer;
+
+import static jalf.DSL.*;
+import static jalf.fixtures.SuppliersAndParts.*;
+import jalf.*;
+
+import org.junit.Test;
+
+public class TestOptimizedUnion extends OptimizerTest {
+
+	@Test
+	public void testProject(){
+		Relation suppliers = suppliers();
+		Relation parts = parts();
+		Relation operator = union(suppliers, parts);
+
+		Relation optimized = null;
+		Relation expected = null;
+
+		// when the three projections are needed
+		// project(
+		// 		union(
+		//			project(supplier, [:NAME, :CITY]), 
+		//			project(shipments, [:NAME, :CITY])),
+		//		[:NAME, :CITY]
+		optimized = optimized(operator).project(attrs(NAME, CITY));
+		expected = project(
+				union(
+						project(suppliers, attrs(NAME, CITY)),
+						project(parts, attrs(NAME, CITY))),
+						attrs(NAME, CITY));
+		assertSameExpression(expected, optimized);
+		
+	}
+	
+	@Test
+	public void testRestrict(){
+		Relation suppliers = suppliers();
+		Relation parts = parts();
+		Relation operator = union(suppliers, parts);
+
+		Relation optimized = null;
+		Relation expected = null;
+
+		// when the restriction applies to left only
+		optimized = optimized(operator).restrict(eq(CITY, "London"));
+		expected = suppliers
+				.restrict(eq(CITY, "London"))
+				.union(parts);
```

Comment:
Sorry because I am not really fresh with this code but is this test correct? To have the same headers, shouldn't `parts` be restricted to CITY as well?